crates/wasmtime/src/runtime/component/store.rs

```diff
@@ -25,5 +27,43 @@ macro_rules! component_store_data {
 
 component_store_data! {
     funcs => crate::component::func::FuncData,
-    instances => Option<Box<crate::component::instance::InstanceData>>,
+    instances => Option<Box<InstanceData>>,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct ComponentInstanceId(usize);
+
+impl ComponentInstanceId {
+    pub fn from_index(idx: usize) -> ComponentInstanceId {
+        ComponentInstanceId(idx)
+    }
+}
+
+impl ComponentStoreData {
+    pub fn next_component_instance_id(&self) -> ComponentInstanceId {
+        ComponentInstanceId(self.instances.len())
+    }
+}
+
+impl StoreData {
+    pub(crate) fn push_component_instance(
+        &mut self,
+        data: Box<InstanceData>,
+    ) -> Stored<Option<Box<InstanceData>>> {
+        assert_eq!(
+            data.instance().id(),
+            self.components.next_component_instance_id()
+        );
+        self.insert(Some(data))
+    }
+}
+
+impl StoreOpaque {
+    #[expect(dead_code, reason = "to be used soon")]
+    pub(crate) fn component_instance(&self, id: ComponentInstanceId) -> &ComponentInstance {
```

Comment:
Will we also be adding a `component_instance_mut` method for getting a `&mut ComponentInstance` from a `&mut self`?

Reply:
Indeed! I'll go ahead and add that in to indicate that it's desired too.